src/daft-local-execution/src/sinks/hash_join_build.rs

```diff
@@ -1,24 +1,61 @@
-use std::sync::Arc;
+use std::sync::{Arc, OnceLock};
 
 use common_error::DaftResult;
+use common_runtime::RuntimeRef;
 use daft_core::prelude::SchemaRef;
 use daft_dsl::ExprRef;
 use daft_logical_plan::JoinType;
 use daft_micropartition::MicroPartition;
 use daft_table::{make_probeable_builder, ProbeState, ProbeableBuilder, Table};
 
-use super::blocking_sink::{BlockingSink, BlockingSinkState, BlockingSinkStatus};
-use crate::pipeline::PipelineResultType;
+use super::blocking_sink::{
+    BlockingSink, BlockingSinkFinalizeResult, BlockingSinkSinkResult, BlockingSinkState,
+    BlockingSinkStatus,
+};
+
+/// ProbeStateBridge is a bridge between the build and probe phase of a hash join.
+/// It is used to pass the probe state from the build phase to the probe phase.
+/// The build phase sets the probe state once building is complete, and the probe phase
+/// waits for the probe state to be set via the `get_probe_state` method.
+pub(crate) type ProbeStateBridgeRef = Arc<ProbeStateBridge>;
+pub(crate) struct ProbeStateBridge {
+    inner: OnceLock<Arc<ProbeState>>,
+    notify: tokio::sync::Notify,
+}
```

Comment:
https://docs.rs/tokio/latest/tokio/sync/struct.OnceCell.html doesn't work cuz there's no `get` method that is async. 

Opted for just a manual implementation.